File: Atcoder/ABC335/C_Loong_Tracking.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    int n,q;
    cin >> n >> q;
    deque<pair<int,int> > que;
    for (int i = n;i>0;i--){
        que.push_back(make_pair(i,0));
    }
    for (int m = 0;m<q;m++){
        int type;
        cin >> type;
        if (type==1){
            // Move the dragon head
            char movement;
            cin >> movement;
            if (movement =='U'){
                que.push_back(make_pair(que.back().first, que.back().second + 1));
            }
            else if (movement == 'D'){
                que.push_back(make_pair(que.back().first, que.back().second - 1));
            }
            else if (movement == 'L'){
                que.push_back(make_pair(que.back().first-1, que.back().second));
            }
            else if (movement == 'R'){
                que.push_back(make_pair(que.back().first+1, que.back().second));
            }
            que.pop_front();
        }
        else {
            //track point
            int pos;
            cin >> pos;
            pair<int, int> coords = que.at(que.size()-pos);
            cout << coords.first << " " << coords.second << endl;
        }
    }
}

File: Atcoder/ABC335/C_Loong_Tracking.py
n,q = map(int, input().split())
from collections import deque
que = deque()
for num in range(n):
    que.append([n-num, 0])
#print(que)
for _ in range(q):
    a = list(map(str, input().split()))
    if a[0] == '1':
        if a[1] == 'U':
            que.append([que[-1][0], que[-1][1]+1])
        elif a[1] == 'D':
            que.append([que[-1][0], que[-1][1]-1])
        elif a[1] == 'L':
            que.append([que[-1][0]-1, que[-1][1]])
        elif a[1] == 'R':
            que.append([que[-1][0]+1, que[-1][1]])
        que.popleft()
    elif a[0] == '2':
        coords = que[-1*int(a[1])]
        print(*coords)